addons/account/wizard/account_payment_register.py

```diff
@@ -456,10 +460,54 @@ def _get_total_amount_in_wizard_currency_to_full_reconcile(self, batch_result):
         for wizard in self:
             if wizard.source_currency_id and wizard.can_edit_wizard:
                 batch_result = wizard._get_batches()[0]
                 wizard.amount = wizard._get_total_amount_in_wizard_currency_to_full_reconcile(batch_result)
+                comp_curr = wizard.company_id.currency_id
+                if wizard.source_currency_id == wizard.currency_id:
+                    # Same currency.
+                    account_move = wizard.line_ids.move_id
+                    has_epd_and_is_in_time = False
+                    if account_move.invoice_payment_term_id.has_early_payment: #TODO apply only on bills ?
+                        has_epd_and_is_in_time = account_move.is_eligible_for_early_discount(wizard.payment_date)
+                    if has_epd_and_is_in_time:
+                        wizard.amount = account_move.early_pay_amount_after_discount
+                        wizard.writeoff_label = _('Early Payment Discount')
+                        invoice_payment_term = wizard.line_ids.move_id.invoice_payment_term_id
+                        wizard.writeoff_account_id = invoice_payment_term.discount_account
+                        wizard.payment_difference_handling = 'reconcile'
+                        total_amount_residual_in_wizard_currency = wizard._get_total_amount_in_wizard_currency_to_full_reconcile(batch_result)
+                        wizard.payment_difference = total_amount_residual_in_wizard_currency - wizard.amount
+                elif wizard.source_currency_id != comp_curr and wizard.currency_id == comp_curr:
+                    # Foreign currency on source line but the company currency one on the opposite line.
+                    wizard.amount = wizard.source_currency_id._convert(
+                        wizard.source_amount_currency,
+                        comp_curr,
+                        wizard.company_id,
+                        wizard.payment_date,
+                    )
+                elif wizard.source_currency_id == comp_curr and wizard.currency_id != comp_curr:
+                    # Company currency on source line but a foreign currency one on the opposite line.
+                    batch = wizard._get_batches()[0]
+                    wizard.amount = abs(sum(
+                        comp_curr._convert(
+                            aml.amount_residual,
+                            wizard.currency_id,
+                            wizard.company_id,
+                            aml.date,
+                        )
+                        for aml in batch['lines']
+                    ))
+                else:
+                    # Foreign currency on payment different than the one set on the journal entries.
+                    wizard.amount = comp_curr._convert(
+                        wizard.source_amount,
+                        wizard.currency_id,
+                        wizard.company_id,
+                        wizard.payment_date,
+                    )
```

Comment:
instead of all this bunch of code, what about making `source_amount_currency` computed based on `line_ids` & `payment_date`. Then, you just need to fill it with the epd amount if eligible